README.md
# Obfuscate Email

A plugin for WordPress that automatically turns plaintext URLs and email addresses into links.

This plugin is available in the WordPress Plugin Directory: https://wordpress.org/plugins/obfuscate-email/


## Installation

1. Install via the built-in WordPress plugin installer. Or install the plugin code inside the plugins directory for your site (typically `/wp-content/plugins/`).
2. Activate the plugin through the 'Plugins' admin menu in WordPress
3. Go to `Settings` -> `Obfuscate Email` admin options page (which you can also get to via the Settings link next to the plugin on the Manage Plugins page) and optionally customize the settings.


## Additional Documentation

See [readme.txt](https://github.com/coffee2code/obfuscate-email/blob/master/readme.txt) for additional usage information. See [CHANGELOG.md](CHANGELOG.md) for the list of changes for each release.


## Support

Commercial support and custom development are not presently available. You can raise an [issue](https://github.com/coffee2code/obfuscate-email/issues) on GitHub or post in the [plugin's support forum on WordPress.org](https://wordpress.org/support/plugin/obfuscate-email/).

If the plugin has been of benefit to you, how about [submitting a review](https://wordpress.org/support/plugin/obfuscate-email/reviews/) for it in the WordPress Plugin Directory or considering a [donation](https://www.paypal.com/cgi-bin/webscr?cmd=_s-xclick&hosted_button_id=6ARCFJ9TX3522)?


## License

This plugin is free software; you can redistribute it and/or modify it under the terms of the [GNU General Public License](https://www.gnu.org/licenses/gpl-2.0.html) as published by the Free Software Foundation; either version 2 of the License, or (at your option) any later version.

CHANGELOG.md
# Changelog

## _(no-progress)_
* Fix: Update plugin framework to 061 to fix a bug preventing settings from getting saved

## 3.8 _(2021-04-10)_

### Highlights:

This minor release updates updates the plugin framework, restructures the unit test file structure, and notes compatibility through WP 5.7+.

### Details:

* Change: Update plugin framework to 060
    * 060:
    * Rename class from `c2c_{PluginName}_Plugin_051` to `c2c_Plugin_060`
    * Move string translation handling into inheriting class making the plugin framework code plugin-agnostic
        * Add abstract function `get_c2c_string()` as a getter for translated strings
        * Replace all existing string usage with calls to `get_c2c_string()`
    * Handle WordPress's deprecation of the use of the term "whitelist"
        * Change: Rename `whitelist_options()` to `allowed_options()`
        * Change: Use `add_allowed_options()` instead of deprecated `add_option_whitelist()` for WP 5.5+
        * Change: Hook `allowed_options` filter instead of deprecated `whitelist_options` for WP 5.5+
    * New: Add initial unit tests (currently just covering `is_wp_version_cmp()` and `get_c2c_string()`)
    * Add `is_wp_version_cmp()` as a utility to compare current WP version against a given WP version
    * Refactor `contextual_help()` to be easier to read, and correct function docblocks
    * Don't translate urlencoded donation email body text
    * Add inline comments for translators to clarify purpose of placeholders
    * Change PHP package name (make it singular)
    * Tweak inline function description
    * Note compatibility through WP 5.7+
    * Update copyright date (2021)
    * 051:
    * Allow setting integer input value to include commas
    * Use `number_format_i18n()` to format integer value within input field
    * Update link to coffee2code.com to be HTTPS
    * Update `readme_url()` to refer to plugin's readme.txt on plugins.svn.wordpress.org
    * Remove defunct line of code
* Change: Move translation of all parent class strings into main plugin file
* Change: Escape markup class attributes before output (for hardening)
* Change: Restructure unit test file structure
    * New: Create new subdirectory `phpunit/` to house all files related to unit testing
    * Change: Move `bin/` to `phpunit/bin/`
    * Change: Move `tests/bootstrap.php` to `phpunit/`
    * Change: Move `tests/` to `phpunit/tests/`
    * Change: Rename `phpunit.xml` to `phpunit.xml.dist` per best practices
* Change: Note compatibility through WP 5.7+
* Change: Update copyright date (2021)

## 3.7 _(2020-06-30)_

### Highlights:

This minor release updates its plugin framework, adds a TODO.md file, updates a few URLs to be HTTPS, expands unit testing, updates compatibility to be WP 4.9 through 5.4+, and minor behind-the-scenes tweaks.

### Details:

* New: Add HTML5 compliance by omitting `type` attribute for `style` tag when the theme supports 'html5'
* Change: Update plugin framework to 050
    * Allow a hash entry to literally have '0' as a value without being entirely omitted when saved
    * Output donation markup using `printf()` rather than using string concatenation
    * Update copyright date (2020)
    * Note compatibility through WP 5.4+
    * Drop compatibility with version of WP older than 4.9
* New: Add TODO.md and move existing TODO list from top of main plugin file into it (and add items to it)
* Change: Note compatibility through WP 5.4+
* Change: Drop compatibility for version of WP older than 4.9
* Change: Update links to coffee2code.com to be HTTPS
* Unit tests:
    * New: Add tests for `add_css()`
    * New: Add test for setting name
    * Change: Store plugin instance in test object to simplify referencing it
    * Change: Update test for default hooks
    * Change: Use HTTPS for link to WP SVN repository in bin script for configuring unit tests (and delete commented-out code)

## 3.6.1 _(2020-01-03)_
* Change: Unit tests: Update unit test install script and bootstrap to use latest WP unit test repo
* Change: Note compatibility through WP 5.3+
* Change: Update copyright date (2020)

## 3.6 _(2019-04-22)_

### Highlights:

This release is a minor update that verifies compatibility through WordPress 5.1+, drops compatibility with versions of WordPress older than 4.7, and makes minor behind-the-scenes improvements.

### Details:

* Change: Initialize plugin on `plugins_loaded` action instead of on load
* Change: Update plugin framework to 049
    * 049:
    * Correct last arg in call to `add_settings_field()` to be an array
    * Wrap help text for settings in `label` instead of `p`
    * Only use `label` for help text for checkboxes, otherwise use `p`
    * Ensure a `textarea` displays as a block to prevent orphaning of subsequent help text
    * Note compatibility through WP 5.1+
    * Update copyright date (2019)
    * 048:
    * When resetting options, delete the option rather than setting it with default values
    * Prevent double "Settings reset" admin notice upon settings reset
    * 047:
    * Don't save default setting values to database on install
    * Change "Cheatin', huh?" error messages to "Something went wrong.", consistent with WP core
    * Note compatibility through WP 4.9+
    * Drop compatibility with version of WP older than 4.7
    * 046:
    * Fix `reset_options()` to reference instance variable `$options`.
    * Note compatibility through WP 4.7+.
    * Update copyright date (2017)
    * 045:
    * Ensure `reset_options()` resets values saved in the database.
* New: Add README.md file
* New: Add CHANGELOG.md file and move all but most recent changelog entries into it
* New: Add inline documentation for hook
* New: Add GitHub link to readme
* Unit tests:
    * New: Add unit test for defaults for settings
    * Change: Improve tests for settings handling
    * Change: Default `WP_TESTS_DIR` to `/tmp/wordpress-tests-lib` rather than erroring out if not defined via environment variable

tests/bootstrap.php
<?php
/**
 * PHPUnit bootstrap file
 *
 * @package Obfuscate_Email
 */

ini_set( 'display_errors', 'on' );
error_reporting( E_ALL );

$_tests_dir = getenv( 'WP_TESTS_DIR' );
if ( ! $_tests_dir ) {
	$_tests_dir = '/tmp/wordpress-tests-lib';
}

// Give access to tests_add_filter() function.
require_once $_tests_dir . '/includes/functions.php';

/**
 * Manually load the plugin being tested.
 */
function _manually_load_plugin() {
	require dirname( dirname( __FILE__ ) ) . '/obfuscate-email.php';
}
tests_add_filter( 'muplugins_loaded', '_manually_load_plugin' );

// Start up the WP testing environment.
require $_tests_dir . '/includes/bootstrap.php';


tests/test-obfuscate-email.php
<?php

defined( 'ABSPATH' ) or die();

class Obfuscate_Email_Test extends WP_UnitTestCase {

	protected $obj;

	public function setUp() {
		parent::setUp();

		$this->obj = c2c_ObfuscateEmail::instance();

		$this->obj->reset_options();
	}


	//
	//
	// DATA PROVIDERS
	//
	//


	public static function get_settings_and_defaults() {
		return array(
			array( 'encode_everything', true ),
			array( 'at_replace', '' ),
			array( 'dot_replace', '' ),
			array( 'use_text_direction', false ),
			array( 'use_display_none', true ),
		);
	}

	public static function get_default_filters() {
		return array(
			array( 'link_description' ),
			array( 'link_notes' ),
			array( 'bloginfo' ),
			array( 'nav_menu_description' ),
			array( 'term_description' ),
			array( 'the_title' ),
			array( 'the_content' ),
			array( 'get_the_excerpt' ),
			array( 'comment_text' ),
			array( 'list_cats' ),
			array( 'widget_text' ),
			array( 'the_author_email' ),
			array( 'get_comment_author_email' ),
		);
	}

	public static function get_default_hooks() {
		$hooks = array(
			array( 'action', 'wp_head', 'add_css' ),
		);

		foreach ( self::get_default_filters() as $filter ) {
			$hooks[] = array( 'filter', $filter[0], 'obfuscate_email', 15 );
		}

		return $hooks;
	}


	//
	//
	// HELPER FUNCTIONS
	//
	//


	private function set_option( $settings = array() ) {
		$defaults = array(
			'encode_everything'  => true,
			'at_replace'         => '',
			'dot_replace'        => '',
			'use_text_direction' => false,
			'use_display_none'   => true,
		);
		$settings = wp_parse_args( $settings, $defaults );
		$this->obj->update_option( $settings, true );
	}


	//
	//
	// TESTS
	//
	//


	public function test_class_exists() {
		$this->assertTrue( class_exists( 'c2c_ObfuscateEmail' ) );
	}

	public function test_get_version() {
		$this->assertEquals( '3.6.1', $this->obj->version() );
	}

	public function test_plugin_framework_class_name() {
		$this->assertTrue( class_exists( 'c2c_ObfuscateEmail_Plugin_050' ) );
	}

	public function test_plugin_framework_version() {
		$this->assertEquals( '050', $this->obj->c2c_plugin_version() );
	}

	public function test_instance_object_is_returned() {
		$this->assertTrue( is_a( $this->obj, 'c2c_ObfuscateEmail' ) );
	}

	public function test_hooks_plugins_loaded() {
		$this->assertEquals( 10, has_action( 'plugins_loaded', array( 'c2c_ObfuscateEmail', 'instance' ) ) );
	}

	/**
	 * @dataProvider get_default_hooks
	 */
	public function test_default_hooks( $hook_type, $hook, $function, $priority = 10, $class_method = true ) {
		$callback = $class_method ? array( $this->obj, $function ) : $function;

		$prio = $hook_type === 'action' ?
			has_action( $hook, $callback ) :
			has_filter( $hook, $callback );

		$this->assertNotFalse( $prio );
		if ( $priority ) {
			$this->assertEquals( $priority, $prio );
		}
	}

	public function test_setting_name() {
		$this->assertEquals( 'c2c_obfuscate_email', c2c_ObfuscateEmail::SETTING_NAME );
	}

	/**
	 * @dataProvider get_default_filters
	 */
	public function test_obfuscation_applies_to_default_filters( $filter ) {
		$email = 'test@example.com';
		$expected = '&#x74;&#x65;&#x73;&#x74;&#x40;<span class="oe_displaynone">null</span>&#x65;&#x78;&#x61;&#x6d;&#x70;&#x6c;&#x65;&#x2e;&#x63;&#x6f;&#x6d;';

		$this->assertEquals( $expected, c2c_obfuscate_email( $email ) );
	}

	/**
	 * @dataProvider get_settings_and_defaults
	 */
	public function test_default_settings( $setting, $value ) {
		$options = $this->obj->get_options();

		if ( is_bool( $value ) ) {
			if ( $value ) {
				$this->assertTrue( $options[ $setting ] );
			} else {
				$this->assertFalse( $options[ $setting ] );
			}
		} else {
			$this->assertEmpty( $options[ $setting ] );
		}
	}

	public function test_at_and_dot_default_replace() {
		$this->set_option( array(
			'encode_everything'  => false,
			'at_replace'         => '',
			'dot_replace'        => '',
			'use_text_direction' => false,
			'use_display_none'   => false,
		) );

		$this->assertEquals( 'test&#064;example&#046;com', c2c_obfuscate_email( 'test@example.com' ) );
	}

	public function test_custom_at_replace() {
		$this->set_option( array(
			'encode_everything'  => false,
			'at_replace'         => 'AT',
			'dot_replace'        => '',
			'use_text_direction' => false,
			'use_display_none'   => false,
		) );

		$this->assertEquals( 'testATexample&#046;com', c2c_obfuscate_email( 'test@example.com' ) );
	}

	public function test_custom_dot_replace() {
		$this->set_option( array(
			'encode_everything'  => false,
			'at_replace'         => '',
			'dot_replace'        => 'DOT',
			'use_text_direction' => false,
			'use_display_none'   => false,
		) );

		$this->assertEquals( 'test&#064;exampleDOTcom', c2c_obfuscate_email( 'test@example.com' ) );
	}

	public function test_text_direction() {
		$this->set_option( array(
			'encode_everything'  => false,
			'at_replace'         => '',
			'dot_replace'        => '',
			'use_text_direction' => true,
			'use_display_none'   => false,
		) );

		$this->assertEquals(
			'<span class="oe_textdirection">moc&#046;elpmaxe&#064;tset</span>',
			c2c_obfuscate_email( 'test@example.com' )
		);
	}

	public function test_display_none() {
		$this->set_option( array(
			'encode_everything'  => false,
			'at_replace'         => '',
			'dot_replace'        => '',
			'use_text_direction' => false,
			'use_display_none'   => true,
		) );

		$this->assertEquals(
			'test&#064;<span class="oe_displaynone">null</span>example&#046;com',
			c2c_obfuscate_email( 'test@example.com' )
		);
	}

	public function test_display_none_and_text_direction_and_at_and_dot_replace() {
		$this->set_option( array(
			'encode_everything'  => false,
			'at_replace'         => 'AT',
			'dot_replace'        => 'DOT',
			'use_text_direction' => true,
			'use_display_none'   => true,
		) );

		$this->assertEquals(
			'<span class="oe_textdirection">mocTODelpmaxe<span class="oe_displaynone">null</span>TAtset</span>',
			c2c_obfuscate_email( 'test@example.com' )
		);
	}

	public function test_encode_everything() {
		$this->set_option( array(
			'encode_everything'  => true,
			'at_replace'         => '',
			'dot_replace'        => '',
			'use_text_direction' => false,
			'use_display_none'   => false,
		) );

		$this->assertEquals(
			'&#x74;&#x65;&#x73;&#x74;&#x40;&#x65;&#x78;&#x61;&#x6d;&#x70;&#x6c;&#x65;&#x2e;&#x63;&#x6f;&#x6d;',
			c2c_obfuscate_email( 'test@example.com' )
		);
	}

	public function test_everything_enabled() {
		$this->set_option( array(
			'encode_everything'  => true,
			'at_replace'         => 'AT',
			'dot_replace'        => 'DOT',
			'use_text_direction' => true,
			'use_display_none'   => true,
		) );

		$this->assertEquals(
			'<span class="oe_textdirection">&#x6d;&#x6f;&#x63;&#x2e;&#x65;&#x6c;&#x70;&#x6d;&#x61;&#x78;&#x65;<span class="oe_displaynone">null</span>&#x40;&#x74;&#x73;&#x65;&#x74;</span>',
			c2c_obfuscate_email( 'test@example.com' )
		);
	}

	/*
	 * add_css()
	 */

	public function test_add_css( $attr = '', $support_html5 = true ) {
		if ( $support_html5 ) {
			add_theme_support( 'html5' );
		}

		$expected = "<style{$attr}>
	span.oe_textdirection { unicode-bidi: bidi-override; direction: rtl; }
	span.oe_displaynone { display: none; }
</style>\n";

		$this->expectOutputRegex( '~^' . preg_quote( $expected ) . '$~', $this->obj->add_css() );
	}

	public function test_add_css_with_no_html5_support() {
		remove_theme_support( 'html5' );

		$this->test_add_css( ' type="text/css"', false );
	}

	/*
	 * Setting handling
	 */

	public function test_does_not_immediately_store_default_settings_in_db() {
		$option_name = c2c_ObfuscateEmail::SETTING_NAME;
		// Get the options just to see if they may get saved.
		$options     = $this->obj->get_options();

		$this->assertFalse( get_option( $option_name ) );
	}

	public function test_uninstall_deletes_option() {
		$option_name = c2c_ObfuscateEmail::SETTING_NAME;
		$options     = $this->obj->get_options();

		// Explicitly set an option to ensure options get saved to the database.
		$this->set_option( array( 'at_replace' => '(AT)' ) );

		$this->assertNotEmpty( $options );
		$this->assertNotFalse( get_option( $option_name ) );

		c2c_ObfuscateEmail::uninstall();

		$this->assertFalse( get_option( $option_name ) );
	}



	/***
	 * ALL ADMIN AREA RELATED TESTS NEED TO FOLLOW THIS FUNCTION
	 ***/



	public function test_turn_on_admin() {
		if ( ! defined( 'WP_ADMIN' ) ) {
			define( 'WP_ADMIN', true );
		}

		// Unhook existing hooks. Then re-register them.
		// Necessary due to the singleton nature of the object and where the
		// is_admin() occurs.
		foreach ( $this->get_default_filters() as $filter ) {
			remove_filter( $filter[0], array( $this->obj, 'obfuscate_email' ), 15 );
		}
		remove_action( 'wp_head', array( $this->obj, 'add_css' ) );

		$this->obj->register_filters();

		$this->assertTrue( is_admin() );
	}

	/**
	 * @dataProvider get_default_filters
	 */
	public function test_does_not_hook_default_filters_in_admin( $filter ) {
		$this->test_turn_on_admin();

		$this->assertFalse( has_filter( $filter, array( $this->obj, 'obfuscate_email' ) ) );
	}

	public function test_nothing_obfuscated_in_admin_even_with_everything_enabled() {
		$this->set_option( array(
			'encode_everything'  => true,
			'at_replace'         => 'AT',
			'dot_replace'        => 'DOT',
			'use_text_direction' => true,
			'use_display_none'   => true,
		) );

		$this->test_turn_on_admin();

		$text = 'test@example.com';

		$this->assertEquals(
			wpautop( $text ),
			apply_filters( 'the_content', $text )
		);
	}

}


obfuscate-email.php
<?php
/**
 * Plugin Name: Obfuscate Email
 * Version:     3.8
 * Plugin URI:  https://coffee2code.com/wp-plugins/obfuscate-email/
 * Author:      Scott Reilly
 * Author URI:  https://coffee2code.com/
 * Text Domain: obfuscate-email
 * License:     GPLv2 or later
 * License URI: https://www.gnu.org/licenses/gpl-2.0.html
 * Description: Obfuscate email addresses to deter email-harvesting spammers.
 *
 * Compatible with WordPress 4.9+ through 5.7+.
 *
 * =>> Read the accompanying readme.txt file for instructions and documentation.
 * =>> Also, visit the plugin's homepage for additional information and updates.
 * =>> Or visit: https://wordpress.org/plugins/obfuscate-email/
 *
 * @package Obfuscate_Email
 * @author  Scott Reilly
 * @version 3.8
 */

/*
	Copyright (c) 2005-2021 by Scott Reilly (aka coffee2code)

	This program is free software; you can redistribute it and/or
	modify it under the terms of the GNU General Public License
	as published by the Free Software Foundation; either version 2
	of the License, or (at your option) any later version.

	This program is distributed in the hope that it will be useful,
	but WITHOUT ANY WARRANTY; without even the implied warranty of
	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
	GNU General Public License for more details.

	You should have received a copy of the GNU General Public License
	along with this program; if not, write to the Free Software
	Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
*/

defined( 'ABSPATH' ) or die();

if ( ! class_exists( 'c2c_ObfuscateEmail' ) ) :

require_once( dirname( __FILE__ ) . DIRECTORY_SEPARATOR . 'c2c-plugin.php' );

final class c2c_ObfuscateEmail extends c2c_Plugin_061 {

	/**
	 * Name of plugin's setting.
	 *
	 * @since 3.6
	 * @var string
	 */
	const SETTING_NAME = 'c2c_obfuscate_email';

	/**
	 * The one true instance.
	 *
	 * @var c2c_ObfuscateEmail
	 */
	private static $instance;

	/**
	 * Default CSS class to indicate text direction enabled (public until filter is added).
	 *
	 * @var string
	 */
	public $css_text_direction  = 'oe_textdirection';

	/**
	 * CSS class to indicate non-display (public until filter is added).
	 *
	 * @var string
	 */
	public $css_display_none    = 'oe_displaynone';

	/**
	 * Get singleton instance.
	 *
	 * @since 3.2
	 */
	public static function instance() {
		if ( ! isset( self::$instance ) ) {
			self::$instance = new self();
		}

		return self::$instance;
	}

	/**
	 * Constructor.
	 */
	protected function __construct() {
		parent::__construct( '3.8', 'obfuscate-email', 'c2c', __FILE__, array() );
		register_activation_hook( __FILE__, array( __CLASS__, 'activation' ) );

		return self::$instance = $this;
	}

	/**
	 * Handles activation tasks, such as registering the uninstall hook.
	 *
	 * @since 3.0
	 */
	public static function activation() {
		register_uninstall_hook( __FILE__, array( __CLASS__, 'uninstall' ) );
	}

	/**
	 * Handles uninstallation tasks, such as deleting plugin options.
	 *
	 * @since 3.0
	 */
	public static function uninstall() {
		delete_option( self::SETTING_NAME );
	}

	/**
	 * Initializes the plugin's config data array.
	 *
	 *
	 * @since 3.0
	 */
	public function load_config() {
		$this->name      = __( 'Obfuscate Email', 'obfuscate-email' );
		$this->menu_name = $this->name;

		$this->config = array(
			'encode_everything' => array(
				'input'   => 'checkbox',
				'default' => true,
				'label'   => __( 'Obfuscate entire email address?', 'obfuscate-email' ),
				'help'    => __( 'All characters of the email address will be obfuscated using hexadecimal HTML entity substitution.', 'obfuscate-email' ),
			),
			'at_replace' => array(
				'input'   => 'text',
				'default' => '',
				'label'   => __( 'Replacement for \'<code>@</code>\'', 'obfuscate-email' ),
				'help'    => __( 'Only applicable if \'Obfuscate entire email address?\' is not checked.<br />Ex. set this to \'[at]\' to get person[at]email.com<br />If applicable but not defined, then <code>&amp;#064;</code> (which is the encoding for &#064;) will be used.', 'obfuscate-email' ),
			),
			'dot_replace' => array(
				'input'   => 'text',
				'default' => '',
				'label'   => __( 'Replacement for \'<code>.</code>\'', 'obfuscate-email' ),
				'help'    => __( 'Only applicable if \'Obfuscate entire email address?\' is not checked.<br />Ex. set this to \'[dot]\' to get person@email[dot]com<br />If applicable but not defined, then <code>&amp;#046;</code> (which is the encoding for &#046;) will be used.', 'obfuscate-email' ),
			),
			'use_text_direction' => array(
				'input'   => 'checkbox',
				'default' => false,
				'label'   => __( 'Utilize CSS text direction technique?', 'obfuscate-email' ),
				'help'    => __( 'This reverses the email address strings as they appear in the markup and utilizes CSS to reverse the text back to the correct direction for visitors to see/use. Note that copying-and-pasting of unlinked text email addresses will result in the pasted text being reversed. Copying the email address (via right-click menu) will work properly.', 'obfuscate-email' ),
			),
			'use_display_none' => array(
				'input'   => 'checkbox',
				'default' => true,
				'label'   => __( 'Utilize CSS display:none technique?', 'obfuscate-email' ),
				'help'    => __( 'This embeds extraneous text within email address strings and then utilizes CSS to hide them so they don\'t appear to visitors.', 'obfuscate-email' ),
			),
		);
	}

	/**
	 * Override the plugin framework's register_filters() to actually hook actions/filters.
	 *
	 * @since 3.0
	 * @since 3.4 Don't do anything if in the admin area.
	 */
	public function register_filters() {
		// Don't obfuscate email addresses in the admin area.
		if ( is_admin() ) {
			return;
		}

		/**
		 * Filters the list of filters processed by the plugin.
		 *
		 * @since 3.0
		 *
		 * @param array $filters The array of filters.
		 */
		$filters = (array) apply_filters(
			'c2c_obfuscate_email_filters',
			array(
				'link_description', 'link_notes', 'bloginfo', 'nav_menu_description',
				'term_description', 'the_title', 'the_content', 'get_the_excerpt', 'comment_text', 'list_cats', 'widget_text',
				'the_author_email', 'get_comment_author_email',
			)
		);

		foreach( $filters as $filter ) {
			add_filter( $filter, array( $this, 'obfuscate_email' ), 15 );
		}

		add_action( 'wp_head', array( $this, 'add_css' ) );
	}

	/**
	 * Returns translated strings used by c2c_Plugin parent class.
	 *
	 * @since 3.8
	 *
	 * @param string $string Optional. The string whose translation should be
	 *                       returned, or an empty string to return all strings.
	 *                       Default ''.
	 * @return string|string[] The translated string, or if a string was provided
	 *                         but a translation was not found then the original
	 *                         string, or an array of all strings if $string is ''.
	 */
	public function get_c2c_string( $string = '' ) {
		$strings = array(
			'A value is required for: "%s"'
				/* translators: %s: Label for setting. */
				=> __( 'A value is required for: "%s"', 'obfuscate-email' ),
			'Click for more help on this plugin'
				=> __( 'Click for more help on this plugin', 'obfuscate-email' ),
			' (especially check out the "Other Notes" tab, if present)'
				=> __( ' (especially check out the "Other Notes" tab, if present)', 'obfuscate-email' ),
			'Coffee fuels my coding.'
				=> __( 'Coffee fuels my coding.', 'obfuscate-email' ),
			'Did you find this plugin useful?'
				=> __( 'Did you find this plugin useful?', 'obfuscate-email' ),
			'Donate'
				=> __( 'Donate', 'obfuscate-email' ),
			'Expected integer value for: %s'
				=> __( 'Expected integer value for: %s', 'obfuscate-email' ),
			'Invalid file specified for C2C_Plugin: %s'
				/* translators: %s: Path to the plugin file. */
				=> __( 'Invalid file specified for C2C_Plugin: %s', 'obfuscate-email' ),
			'More information about %1$s %2$s'
				/* translators: 1: plugin name 2: plugin version */
				=> __( 'More information about %1$s %2$s', 'obfuscate-email' ),
			'More Help'
				=> __( 'More Help', 'obfuscate-email' ),
			'More Plugin Help'
				=> __( 'More Plugin Help', 'obfuscate-email' ),
			'Please consider a donation'
				=> __( 'Please consider a donation', 'obfuscate-email' ),
			'Reset Settings'
				=> __( 'Reset Settings', 'obfuscate-email' ),
			'Save Changes'
				=> __( 'Save Changes', 'obfuscate-email' ),
			'See the "Help" link to the top-right of the page for more help.'
				=> __( 'See the "Help" link to the top-right of the page for more help.', 'obfuscate-email' ),
			'Settings'
				=> __( 'Settings', 'obfuscate-email' ),
			'Settings reset.'
				=> __( 'Settings reset.', 'obfuscate-email' ),
			'Something went wrong.'
				=> __( 'Something went wrong.', 'obfuscate-email' ),
			'The plugin author homepage.'
				=> __( 'The plugin author homepage.', 'obfuscate-email' ),
			"The plugin configuration option '%s' must be supplied."
				/* translators: %s: The setting configuration key name. */
				=>__( "The plugin configuration option '%s' must be supplied.", 'obfuscate-email' ),
			'This plugin brought to you by %s.'
				/* translators: %s: Link to plugin author's homepage. */
				=> __( 'This plugin brought to you by %s.', 'obfuscate-email' ),
		);

		if ( ! $string ) {
			return array_values( $strings );
		}

		return ! empty( $strings[ $string ] ) ? $strings[ $string ] : $string;
	}

	/**
	 * Output CSS.
	 */
	public function add_css() {
		$type_attr = current_theme_supports( 'html5' ) ? '' : ' type="text/css"';

echo <<<HTML
<style{$type_attr}>
	span.{$this->css_text_direction} { unicode-bidi: bidi-override; direction: rtl; }
	span.{$this->css_display_none} { display: none; }
</style>

HTML;
	}

	/**
	 * Obfuscate plaintext emails.
	 *
	 * @param string  $text The text containing emails to obfuscate.
	 * @param array   $args An array of configuration options, each element of which will override the plugin's corresponding default setting.
	 * @return string The text with emails obfuscated.
	 */
	public function obfuscate_email( $text, $args = array() ) {
		$options = $this->get_options();

		if ( $args ) {
			$options = $this->options = wp_parse_args( $args, $options );
		}

		$text = ' ' . $text . ' ';

		$cb               = array( $this, 'obfuscate_email_cb' );
		$cb_for_attribute = array( $this, 'obfuscate_email_in_attribute_cb' );

		// pre-3.0 regex : "#(([a-z0-9\-_\.]+?)@([^\s,{}\(\)\[\]]+\.[^\s.,{}\(\)\[\]]+))#iesU"

		// This isn't making any attempts to only recognize valid email address
		// syntax. Basically anything roughly like x@y.zz looks like an email address.
		$email_regex = '([.0-9a-z_+-]+)@(([0-9a-z-]+\.)+[0-9a-z]{2,})';

		// This matches emails except for those appearing in tag attributes.
		// (only need that refinement if wrapping text in tags)
		if ( $options['use_text_direction'] || $options['use_display_none'] ) {
			$text = preg_replace_callback( "#(?!<.*?)$email_regex(?![^<>]*?>)#i", $cb, $text );
		}

		// If checking again, then the only emails that are left are those in
		// attributes. Is the first check if use_text_direction is false.
		$text = preg_replace_callback( "#$email_regex#i", $cb_for_attribute, $text );

		return trim( $text );
	}

	/**
	 * Obfuscate plaintext emails callback for email address appearing in tag
	 * attribute.
	 *
	 * @since 3.1
	 *
	 * @param string  $email The email to obfuscate.
	 * @return string The obfuscated email.
	 */
	public function obfuscate_email_in_attribute_cb( $matches ) {
		return $this->obfuscate_email_cb( $matches, false );
	}

	/**
	 * Obfuscate plaintext emails callback.
	 *
	 * @param string  $email The email to obfuscate.
	 * @return string The obfuscated email.
	 */
	public function obfuscate_email_cb( $matches, $allow_markup = true ) {
		$options = $this->get_options();

		$email_name = $matches[1];
		$email_host = $matches[2];

		if ( $allow_markup && $options['use_text_direction'] ) {
			$email_name = strrev( $email_name );
			$email_host = strrev( $email_host );
		}

		if ( $options['encode_everything'] ) {
			$email_name = substr( chunk_split( bin2hex( " $email_name" ), 2, ";&#x" ), 3, -3 );
			$email_host = substr( chunk_split( bin2hex( " $email_host" ), 2, ";&#x" ), 3, -3 );
			$at = '&#x40;';
		} else {
			$at  = empty( $options['at_replace'] )  ? '&#064;' : $options['at_replace'];
			$dot = empty( $options['dot_replace'] ) ? '&#046;' : $options['dot_replace'];

			if ( ! $allow_markup ) {
				$at  = esc_attr( trim( $at ) );
				$dot = esc_attr( trim( $dot ) );
			}

			// If reversing text, also reverse the AT and DOT replacements unless they
			// look like they are HTML encoded entities. Very crude.
			if ( $options['use_text_direction'] ) {
				if ( false === strpos( $at, '&' ) ) {
					$at = strrev( $at );
				}
				if ( false === strpos( $dot, '&' ) ) {
					$dot = strrev( $dot );
				}
			}

			$email_host = str_replace( '.', $dot, $email_host );
		}

		if ( $allow_markup && $options['use_display_none'] ) {
			$none = '<span class="' . esc_attr( $this->css_display_none ) . '">null</span>';
			if ( $options['use_text_direction'] ) {
				$at = $none . $at;
			} else {
				$at .= $none;
			}
		}

		if ( $allow_markup && $options['use_text_direction'] ) {
			$email = $email_host . $at . $email_name;
		} else {
			$email = $email_name . $at . $email_host;
		}

		if ( $allow_markup && $options['use_text_direction'] ) {
			$email = '<span class="' . esc_attr( $this->css_text_direction ) . '">' . $email . '</span>';
		}

		return $email;
	}

} // end c2c_ObfuscateEmail

add_action( 'plugins_loaded', array( 'c2c_ObfuscateEmail', 'instance' ) );

endif; // end if !class_exists()


if ( ! function_exists( 'c2c_obfuscate_email' ) ) :
/**
 * Text to parse for plaintext emails to be obfuscated.
 *
 * @param string  $text  The text to parse for email addresses to obfuscate.
 * @param array   $args  An array of configuration options, each element of which will override the plugin's corresponding default setting.
 * @return string The text with email addresses obfuscated.
 */
function c2c_obfuscate_email( $text, $args = array(), $deprecated_1 = null, $deprecated_2 = null, $deprecated_3 = null ) {
	if ( ! is_array( $args ) || $deprecated_1 || $deprecated_2 || $deprecated_3 ) {
		_deprecated_argument( __FUNCTION__, '3.0', 'Use $args array argument to override default settings.' );
	}

	return c2c_ObfuscateEmail::instance()->obfuscate_email( $text, $args );
}
endif;


/**
 * *******************
 * DEPRECATED FUNCTIONS
 * *******************
 */

	// To be removed in v3.1 of plugin
	if ( ! function_exists( 'c2c_email_obfuscator' ) ) :
	/**
	 * Obfuscate plaintext emails.
	 *
	 * @deprecated since 3.0 Use c2c_obfuscate_email()
	 *
	 * @param string  $email The email to obfuscate.
	 * @return string The obfuscated email.
	 */
	function c2c_email_obfuscator( $email, $encode_everything = true, $at_replace = '&#064;', $dot_replace = '&#046;' ) {
		_deprecated_function( __FUNCTION__, '3.0', 'c2c_obfuscate_email' );
		return c2c_obfuscate_email( $email );
	}
	endif;
